### PR TITLE
Upload to NFT storage

### LINKED ISSUE #68: Deploy with IPFS
Now one other issue is how all these different deployed contracts would tie together and this would be through the factory ID (which we currently allow to be passed when the SDK is initialised.

Currently the `deploy` function just accepts any IPFS URL...

```tsx
export async function deploy({
  signer,
  nft,
}: {
  signer: Signer;
  nft: NFTMetadata;
}) {
  const openFormatContract = new ethers.ContractFactory(
    base.abi,
    base.bytecode,
    signer
  );

  const contract = await openFormatContract.deploy(
    nft.name,
    nft.symbol,
    nft.url,
    nft.maxSupply,
    ethers.utils.parseEther(nft.mintingPrice.toString())
  );

  const receipt = await contract.deployTransaction.wait();

  return receipt;
}
```

But this _really_ should not be the case. We should really be handling the way the factory ID is stored.

I propose instead of passing through the URL, we pass through any metadata (optionally). So this...

```tsx
await sdk.deploy({
  maxSupply: 100,
  mintingPrice: 0.01,
  name: 'Test',
  symbol: 'TEST',
  url: 'ipfs://',
});
```

...would becomes this...

```tsx
await sdk.deploy({
  maxSupply: 100,
  mintingPrice: 0.01,
  name: 'Test',
  symbol: 'TEST',
  metadata: { },
});
```

Internally inside the SDKs `deploy` function we should add the factory ID into this metadata, upload it to IPFS and then handle getting that URL out.

### PR DESCRIPTION
**Pull request by benjaminreid**
This aims to fix #68 by handling the upload of the IPFS data to https://nft.storage/.

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Code complete
- [x] Changeset

### CODE REVIEW

**Comment by benjaminreid** [comment]
There's a few things to clear up before this can be wrapped up. I'll try explain what this PR does.

Currently, you can pass your own `url` to `deploy`. This assumes you've uploaded your NFT with relevant metadata and got the URL back. This PR does not change this, it attempts to avoid a breaking change.

```tsx
const sdk = new OpenFormatSDK({
  network: 'http://localhost:8545',
  signer,
});

const receipt = await sdk.deploy({
  name: 'Test',
  url: 'ipfs://',
  symbol: 'TEST',
  maxSupply: 100,
  mintingPrice: 0.01,
});
```

Here's where things start to change. If you omit a `url`, it will throw an error when trying to deploy.

```tsx
const sdk = new OpenFormatSDK({
  network: 'http://localhost:8545',
  signer,
});

const receipt = await sdk.deploy({
  name: 'Test',
  symbol: 'TEST',
  maxSupply: 100,
  mintingPrice: 0.01,
});
```

However. If you specify a `factory` and omit a `url`, it will attempt to upload your NFT for you via nft.storage.

```tsx
const sdk = new OpenFormatSDK({
  network: 'http://localhost:8545',
  signer,
  factory: "some-id",
});

const receipt = await sdk.deploy({
  name: 'Test',
  symbol: 'TEST',
  maxSupply: 100,
  mintingPrice: 0.01,
});
```

Now uploading the NFT is really nice on behalf of the users, though there's some things worth double checking in this PR.

If you specify a `factory` and omit the `url` (like above) it will now throw an error if you don't supply some assumed mandatory fields. You now need to additionally pass `description`, `image` and `releaseType`.

`description` and `image` are required fields of NFT storage and I've made `releaseType` required as it seems to make sense when the metadata is built with `factory_id`.

```tsx
const sdk = new OpenFormatSDK({
  network: 'http://localhost:8545',
  signer,
  factory: "some-id",
});

const receipt = await sdk.deploy({
  name: 'Test',
  description: "some description",
  image: new Blob([""], {type : 'image/png'}),
  releaseType: "image",
  symbol: 'TEST',
  maxSupply: 100,
  mintingPrice: 0.01,
});
```

You can also optionally supply an object of metadata which can include other files.

```tsx
const receipt = await sdk.deploy({
  name: 'Test',
  description: "some description",
  image: new Blob([""], {type : 'image/png'}),
  releaseType: "image",
  symbol: 'TEST',
  maxSupply: 100,
  mintingPrice: 0.01,
  metadata: {
    some: 'test',
    audio: new File(...),
  },
});
```

Is sounds a bit confusing given there a various optional things now but really this is the main change to the deploy function...

```tsx
// an optional URL a user can supply when calling the deploy function
let url = nft.url;

// if the url is not provided, but they've set factory ID when intialising the SDK, we attempt to upload the NFT for them
if (typeof url === 'undefined' && factory) {
  // we check to see if a few required fields have been passed, these are optional in TypeScript due to being able to pass your own url
  invariant(nft.description, 'A description must be set');
  invariant(nft.image, 'An image must be set');
  invariant(nft.releaseType, 'A release type must be set');

  const customMetadata = nft.metadata ?? {};

  // we send the NFT to storage, spreading any other metadata
  const metadata = await nftStorage.store({
    name: nft.name,
    description: nft.description,
    image: nft.image,
    factory_id: factory,
    release_type: nft.releaseType,
    ...customMetadata,
  });

  url = metadata.url;
}

// is no url has been supplied and no factory id was set to perform an upload, an error is thrown
invariant(url, 'An IPFS URL must be set');
```

One thing worth checking with the nft.storage api is that in most of our examples that have been built so far, the metadata is spread across the root stored object.

```tsx
await nftStorage.store({
  name: nft.name,
  description: nft.description,
  image: nft.image,
  // custom data
  factory_id: factory,
  release_type: nft.releaseType,
});
```

Whereas in their docs, they specify a `properties` field should be used. https://nftstorage.github.io/nft.storage/client/classes/lib.NFTStorage.html#store

So I'm not sure if we need to change things to... in our code.

```tsx
const metadata = await nftStorage.store({
  name: nft.name,
  description: nft.description,
  image: nft.image,
  properties: {
    factory_id: factory,
    release_type: nft.releaseType,
    ...customMetadata,
  },
});
```

**Comment by benjaminreid** [comment]
Ideas from @coatezy: Potential for the factory ID to be optional and potential for the factory ID to be passed through with the deploy function rather than when initialising the SDK.

**Comment by RuariM** [comment]
The `properties`  field is a metadata storage standard that some NFT marketplaces use. For example Opensea will automatically detect metadata if it is stored in key value pairs inside the `properties` field.

**Comment by benjaminreid** [comment]
> The `properties` field is a metadata storage standard that some NFT marketplaces use. For example Opensea will automatically detect metadata if it is stored in key value pairs inside the `properties` field.

That makes sense, thanks. I guess we can ignore that for now then and keep the data how we currently have it.

**Comment by RuariM** [comment]
cool, for reference they talk about it here - https://nft.storage/docs/how-to/mint-erc-1155/

_'When adding custom properties, we recommend that you put them in the properties object, rather than at the top level. However, if you're minting for a specific marketplace, you should consult their documentation and follow their recommendations.'_

**Comment by RuariM** [comment]
Moving my changes into a separate PR.

**Comment by Sam-Archie** [comment]
Is it worth refactoring open-format-NFT to also use the properties standard?? I think that whatever route we go down are examples should be inline.

**Comment by benjaminreid** [comment]
> @benjaminreid Is it worth refactoring open-format-NFT to also use the properties standard?? I think that whatever route we go down are examples should be inline.

The properties weren't used for the subgraph stuff in the end, so should be okay to stay as is 👍

**Comment by tinypell3ts** [comment]
Great work on this. I think it's important to note in the documentation that the metadata can currently only be a key/value pair with the value being a string. Please see [here](https://github.com/simpleweb/open-format-subgraph/blob/main/src/mappings/mapping-v1.3.0.ts#L153-L176) for more. Nested metadata is on the roadmap for future releases.

**Comment by benjaminreid** [comment]
@tinypell3ts Hopefully this note covers it https://open-format-docs-pahg5ufeh-simpleweb-ltd.vercel.app/guides/deploying

File: sdks/open-format/src/core/contract.ts [hunk 1]
```diff
@@ -419,27 +421,64 @@ export async function burn({
...
 }) {
   const openFormatContract = new ethers.ContractFactory(
     base.abi,
     base.bytecode,
     signer
   );
 
+  let url = nft.url;
+
+  if (typeof url === 'undefined' && factory) {
```
Comment (tinypell3ts) [inline sdks/open-format/src/core/contract.ts:449]: factory should be optional here.

File: docs/docs/guides/deploying.mdx [hunk 2]
```diff
@@ -3,21 +3,112 @@ import TabItem from '@theme/TabItem';
...
+  network: 'mumbai',
+  signer: '0x...',
+  factory: 'YOUR-FACTORY-ID',
+  nftStorageToken: 'xxx',
+});
+
+await sdk.deploy({
+  maxSupply: 100,
+  mintingPrice: 0.01,
+  name: "TICKET",
```
Comment (tinypell3ts) [inline docs/docs/guides/deploying.mdx:36]: Can we add description and image here as they are required.

File: docs/docs/guides/deploying.mdx [hunk 3]
```diff
@@ -3,21 +3,112 @@ import TabItem from '@theme/TabItem';
...
+  symbol: "FESTIVAL",
+  releaseType: "ticket",
+  metadata: {
+    location: "Glastonbury",
```
Comment (tinypell3ts) [inline docs/docs/guides/deploying.mdx:40]: Can we add a metadata type key/string pairs

File: docs/docs/guides/deploying.mdx [hunk 4]
```diff
@@ -3,21 +3,112 @@ import TabItem from '@theme/TabItem';
...
+  nftStorageToken: 'xxx',
+});
+
+await sdk.deploy({
+  maxSupply: 100,
+  mintingPrice: 0.01,
+  name: "TICKET",
+  symbol: "FESTIVAL",
+  releaseType: "ticket",
+  metadata: {
```
Comment (tinypell3ts) [inline docs/docs/guides/deploying.mdx:39]: Can we add an attribute type which is an array of objects (key, value)

File: docs/docs/guides/deploying.mdx [hunk 5]
```diff
@@ -1,44 +1,139 @@
...
+  const { deploy } = useDeploy();
+
+  return (
+    <>
+      <button
+        onClick={() =>
+          deploy({
+            maxSupply: 100,
+            mintingPrice: 0.01,
+            name: "TICKET",
```
Comment (tinypell3ts) [inline docs/docs/guides/deploying.mdx:83]: Please can you add description, image and any another required fields here.

File: docs/docs/guides/deploying.mdx [hunk 6]
```diff
@@ -1,44 +1,139 @@
...
+import { OpenFormatSDK } from "@simpleweb/open-format";
+
+const sdk = new OpenFormatSDK({
+  network: "mumbai",
+  signer: "0x...",
 });
 
 await sdk.deploy({
   maxSupply: 100,
   mintingPrice: 0.01,
```
Comment (tinypell3ts) [inline docs/docs/guides/deploying.mdx:125]: Please can you add description, image and any another required fields here.

File: docs/docs/guides/deploying.mdx [hunk 7]
```diff
@@ -52,7 +147,7 @@ function MyComponent() {
             mintingPrice: 0.01,
             name: "TICKET",
```
Comment (tinypell3ts) [inline docs/docs/guides/deploying.mdx:148]: Please can you add description, image and any another required fields here.

**Review by tinypell3ts — APPROVED** [review]

**Pull request merged**